### ISSUE TITLE
Packet forwarder not working

### ISSUE THREAD

**Issue by uros76** [NebraLtd/helium-rak#16]
Hi. I have forked the fleet and added my MNTD blackspot hotspot to the dashboard. However, the miner doesn't seem to be working. 

Are there any specific variable settings I need to make? 

I get this error:  
![image](https://user-images.githubusercontent.com/25409772/187390578-61557d95-b153-4ad3-ab67-2f114331ed21.png)

**Comment by pritamghanghas** [comment]
you need a VARIANT env variable to be defined to match your hardware. Take
a look at hardware_definitions.py in hm-pyhelper.

Regards,
Nebra

**Comment by uros76** [comment]
Okay, found the value matching RAK/MNTD and now put it as device variable. Now I get this is log... keyerror: variant

Not sure is this correct way? 
![image](https://user-images.githubusercontent.com/25409772/187436475-51b16f89-7552-4643-9494-7f59594d180c.png)

**Comment by Turb0n00by** [comment]
Name 	Value 	Service 	Device overrides 	Actions 
VARIANT	
COMP-RAKHM
All services

**Comment by uros76** [comment]
Awesome! This helped and now it seems the diagnostics container is working. 

However, I see now gateway-config container struggling to work. I rebooted it but still see this error in the live log window... _RuntimeError: Failed to add edge detection_
Is there some additional variable I need to set? 

![image](https://user-images.githubusercontent.com/25409772/187442114-c1e48a7d-b958-45cc-b8b6-badda1cf8aaf.png)

I can't even connect to it:
![image](https://user-images.githubusercontent.com/25409772/187442608-9219a43e-dfa2-4b51-a3bd-8ab24ac4dd1d.png)

**Comment by Turb0n00by** [comment]
Nah you're ok there. I thought that was an error sometime ago but you
should be ok. Expect a day or so for everything to "settle" The dashboard
is a great monitoring tool.

**Comment by uros76** [comment]
Okay will let it work and hope that gateway-config survives :) 
Thanks again for help

**Comment by uros76** [comment]
Hi. I let it run since yesterday and I think there is still something wrong with gateway-config. Device diagnostics on the dashboard says it's failed and restarted many times... 
![image](https://user-images.githubusercontent.com/25409772/187622617-12e1a181-83b6-4b48-a51c-345e16ecdeb1.png)

Any ideas how to get it working? Or this is not really needed?

**Comment by VeniceInventors** [comment]
Same issue here.

I added the variable and keep getting the `gateway-config  RuntimeError: Failed to add edge detection` error in the log, which then causes the gateway-config service to go in a "stop and restart" loop.

Also packet-forwarder stops and restarts with the error:
 json.decoder.JSONDecodeError: Expecting property name enclosed in double quotes: line 7 column 28 (char 168)

Here these issues only happen with the image from https://hub.balena.io/organizations/nebraltd/fleets/HELIUM-RAK forked to my own fleet on Balena.
When using the (seemingly*) same image from the above URL without forking it, it seems to be fine, or at least the local diagnostic page shows the radio as being operational so the packet forwarder is working, but that's all I know as I can't access the Balena dashboard to see the logs (not sure where to pull the logs from via ssh).
[edit] found the logs in 
/mnt/data/docker/volumes/1914219_miner-log/_data/
and
/mnt/data/docker/volumes/1914219_watchdog-log/_data/
[/edit]

*My understanding is that only the config.json differs between my forked 
balena-cloud-nebra-rak-raspberrypi4-64-2.105.1+rev1-v14.2.0.img
and the original
balena-cloud-HELIUM-RAK-raspberrypi4-64-2.105.1+rev1-v14.2.0.img
so my guess is there must be some additional settings or variables on the HELIUM-RAK fleet that would be missing in the forked fleet, like that VARIANT variable.

**Comment by uros76** [comment]
I have this set on the fleet and it works despite endless gateway-config errors
![image](https://user-images.githubusercontent.com/25409772/194081527-8a5bb92e-1e0c-41a4-bd73-9736079944ca.png)

**Comment by VeniceInventors** [comment]
> I have this set on the fleet and it works despite endless gateway-config errors
Are you using the *-raspberrypi4-64-2.105.1+rev1-v14.2.0.img image?

My understanding is that the gateway-config is only used access to the diagnostics and wifi config via bluetooth so the hotspot would still be able to witness and transfer data. however with the current image it looks like the packet forwarder is broken too so the radio is down. https://github.com/NebraLtd/hm-pktfwd/issues/108

**Comment by uros76** [comment]
Ok, that's new for me. 
I have my fleet on this 
![image](https://user-images.githubusercontent.com/25409772/194285594-5d6cc6a1-edf1-4f9e-b04f-546f8aafdf98.png)

**Comment by VeniceInventors** [comment]
Ah yes that's why, your screenshot shows firmware 2022.08.17.5, while the latest is 2022.08.17.6-2 but the last working one appears to be 2022.08.17.6

**Comment by uros76** [comment]
> Ah yes that's why, your screenshot shows firmware 2022.08.17.5, while the latest is 2022.08.17.6-2 but the last working one appears to be 2022.08.17.6

That is exactly the reason I am letting mine keep running on 2022.08.17.5. They work just fine with the previous fw. 
My fleet on balena dashboard has this old release and I can add future release to existing fleet anytime. 
I wish there would be a way to share my old forked release with you?

**Comment by Turb0n00by** [comment]
> > Ah yes that's why, your screenshot shows firmware 2022.08.17.5, while the latest is 2022.08.17.6-2 but the last working one appears to be 2022.08.17.6
> 
> That is exactly the reason I am letting mine keep running on 2022.08.17.5. They work just fine with the previous fw. My fleet on balena dashboard has this old release and I can add future release to existing fleet anytime. I wish there would be a way to share my old forked release with you?

You can revert to older images assuming you have ran/forked them in the past. 

Select your device 
<img width="595" alt="Screenshot 2022-10-07 at 6 08 19 AM" src="https://user-images.githubusercontent.com/51140610/194541355-b0a55568-6a46-406f-add2-b77be5d7e717.png">

Select Pin to release
<img width="342" alt="Screenshot 2022-10-07 at 6 08 24 AM" src="https://user-images.githubusercontent.com/51140610/194541389-292044cd-43ed-472d-abed-1292417d8ae2.png">

Select version
<img width="707" alt="Screenshot 2022-10-07 at 6 08 31 AM" src="https://user-images.githubusercontent.com/51140610/194541560-fc9e5c95-89f7-47e7-84e1-87e6edf682cb.png">

**Comment by VeniceInventors** [comment]
Good to know, unfortunately I only forked the current fleet release so the menu only shows that one.

However it's possible to switch the GitHub repo URL when forking the nebra-rak fleet
![image](https://user-images.githubusercontent.com/95456835/194550395-14231458-71fc-44d1-a123-eecdbf4872b8.png)

so I've forked their GitHub repository and edited the docker-compose.yml, to make it match the sept. 9 version from their repo's history. 
I'll try that to see what happens.
[edit]
That didn't work, I'm still getting FIRMWARE_VERSION=2022.08.17.6-2
Either balena built the image from the parent repo instead of the forked one, or the hotspot automatically updated itself to the latest firmware after booting.
[/edit]
[edit2]
Got the packet forwarder working with the previous release. I do get the gateway-config errors but that issue seems to have been there in release 2022.08.17.5 already.
I'll try to find the cause,
[/edit2]

**Comment by Turb0n00by** [comment]
Your variable should read as: 
COMP-RAKHM

**Comment by shawaj** [comment]
@uros76 do you still get these gateway config errors on this RAK firmware?

**Comment by uros76** [comment]
> @uros76 do you still get these gateway config errors on this RAK firmware?

No, once I setup correct variable on fleet it works fine. I also recently updated fleet with latest release, all works ok.

**Comment by shawaj** [comment]
Awesome, will close this issue then 🙂

**Comment by VeniceInventors** [comment]
> Awesome, will close this issue then slightly_smiling_face

The packet forwarder is working but the gateway config error remains here with RAK v2 (mntd blackspot). It keeps causing the python script to abort and restart.
The gpio pin number used for the side button seems to be cause of the problem, or something else is preventing the GPIO.add_event_detect() function from succeeding. See the gateway-config log below.
I could be wrong but it seems the button is connected to the RAK2003 Pi HAT's MCU rather than one of the GPIO pins.

```
22.04.23 04:34:12 (-0700)  gateway-config  Waiting to connect to bluetoothd..
Agent registered
[CHG] Controller E4:5F:01:B2:FD:12 Pairable: yes
22.04.23 04:34:12 (-0700)  gateway-config  [bluetooth]# pairable off
22.04.23 04:34:12 (-0700)  gateway-config  [bluetooth]# quit
FREQ variable is already set as N/A (Non Nebra Miner).
22.04.23 04:34:12 (-0700)  gateway-config  VARIANT variable is already set as COMP-RAKHM.
22.04.23 04:34:12 (-0700)  gateway-config  All required environment variables have been set.
22.04.23 04:34:12 (-0700)  gateway-config  method return time=1682163252.036938 sender=org.freedesktop.DBus -> destination=:1.127 serial=3 reply_serial=2
22.04.23 04:34:12 (-0700)  gateway-config     array [
22.04.23 04:34:12 (-0700)  gateway-config        string "org.freedesktop.DBus"
22.04.23 04:34:12 (-0700)  gateway-config        string ":1.127"
22.04.23 04:34:12 (-0700)  gateway-config     ]
22.04.23 04:34:12 (-0700)  gateway-config  DBus is now accepting connections
22.04.23 04:34:12 (-0700)  gateway-config  libmraa[11]: libmraa version v2.1.0-24-gbd91964 initialised by user 'root' with EUID 0
22.04.23 04:34:12 (-0700)  gateway-config  libmraa[11]: gpio: platform doesn't support chardev, falling back to sysfs
22.04.23 04:34:12 (-0700)  gateway-config  libmraa[11]: libmraa initialised for platform 'Raspberry Pi Model B Rev 1' of type 5
22.04.23 04:34:13 (-0700)  gateway-config  2023-04-22 11:34:13,213 - [DEBUG] - __main__ - (__main__.py).validate_env -- /opt/gatewayconfig/__main__.py:(34) - Starting with the following ENV:
22.04.23 04:34:13 (-0700)  gateway-config          SENTRY_DSN=None
22.04.23 04:34:13 (-0700)  gateway-config          BALENA_APP_NAME=balenamntd
22.04.23 04:34:13 (-0700)  gateway-config          BALENA_DEVICE_UUID=e62a463a5d15eeecf078862dd4bb0823
22.04.23 04:34:13 (-0700)  gateway-config          VARIANT=COMP-RAKHM
22.04.23 04:34:13 (-0700)  gateway-config          ETH0_MAC_ADDRESS_FILEPATH=/sys/class/net/eth0/address
22.04.23 04:34:13 (-0700)  gateway-config          WLAN0_MAC_ADDRESS_FILEPATH=/sys/class/net/wlan0/address
22.04.23 04:34:13 (-0700)  gateway-config          DIAGNOSTICS_JSON_URL=http://diagnostics/json
22.04.23 04:34:13 (-0700)  gateway-config          ETHERNET_IS_ONLINE_FILEPATH=/sys/class/net/eth0/carrier
22.04.23 04:34:13 (-0700)  gateway-config          FIRMWARE_VERSION=v1.0.2-3
22.04.23 04:34:13 (-0700)  gateway-config  
22.04.23 04:34:13 (-0700)  gateway-config  Traceback (most recent call last):
22.04.23 04:34:13 (-0700)  gateway-config    File "/usr/lib/python3.9/runpy.py", line 197, in _run_module_as_main
22.04.23 04:34:13 (-0700)  gateway-config      return _run_code(code, main_globals, None,
22.04.23 04:34:13 (-0700)  gateway-config    File "/usr/lib/python3.9/runpy.py", line 87, in _run_code
22.04.23 04:34:13 (-0700)  gateway-config      exec(code, run_globals)
22.04.23 04:34:13 (-0700)  gateway-config    File "/opt/gatewayconfig/__main__.py", line 77, in <module>
22.04.23 04:34:13 (-0700)  gateway-config      main()
22.04.23 04:34:13 (-0700)  gateway-config    File "/opt/gatewayconfig/__main__.py", line 30, in main
22.04.23 04:34:13 (-0700)  gateway-config      start()
22.04.23 04:34:13 (-0700)  gateway-config    File "/opt/gatewayconfig/__main__.py", line 57, in start
22.04.23 04:34:13 (-0700)  gateway-config      config_app = GatewayconfigApp(
22.04.23 04:34:13 (-0700)  gateway-config    File "/opt/gatewayconfig/gatewayconfig_app.py", line 43, in __init__
22.04.23 04:34:13 (-0700)  gateway-config      self.init_gpio(variant)
22.04.23 04:34:13 (-0700)  gateway-config    File "/opt/gatewayconfig/gatewayconfig_app.py", line 136, in init_gpio
22.04.23 04:34:13 (-0700)  gateway-config      self.user_button = Button(btn, hold_time=USER_BUTTON_HOLD_SECONDS)
22.04.23 04:34:13 (-0700)  gateway-config    File "/opt/venv/lib/python3.9/site-packages/gpiozero/devices.py", line 108, in __call__
22.04.23 04:34:13 (-0700)  gateway-config      self = super(GPIOMeta, cls).__call__(*args, **kwargs)
22.04.23 04:34:13 (-0700)  gateway-config    File "/opt/venv/lib/python3.9/site-packages/gpiozero/input_devices.py", line 414, in __init__
22.04.23 04:34:13 (-0700)  gateway-config      super(Button, self).__init__(
22.04.23 04:34:13 (-0700)  gateway-config    File "/opt/venv/lib/python3.9/site-packages/gpiozero/mixins.py", line 433, in __init__
22.04.23 04:34:13 (-0700)  gateway-config      super(HoldMixin, self).__init__(*args, **kwargs)
22.04.23 04:34:13 (-0700)  gateway-config    File "/opt/venv/lib/python3.9/site-packages/gpiozero/input_devices.py", line 171, in __init__
22.04.23 04:34:13 (-0700)  gateway-config      self.pin.when_changed = self._pin_changed
22.04.23 04:34:13 (-0700)  gateway-config    File "/opt/venv/lib/python3.9/site-packages/gpiozero/pins/__init__.py", line 432, in <lambda>
22.04.23 04:34:13 (-0700)  gateway-config      lambda self, value: self._set_when_changed(value),
22.04.23 04:34:13 (-0700)  gateway-config    File "/opt/venv/lib/python3.9/site-packages/gpiozero/pins/pi.py", line 319, in _set_when_changed
22.04.23 04:34:13 (-0700)  gateway-config      self._enable_event_detect()
22.04.23 04:34:13 (-0700)  gateway-config    File "/opt/venv/lib/python3.9/site-packages/gpiozero/pins/rpigpio.py", line 226, in _enable_event_detect
22.04.23 04:34:13 (-0700)  gateway-config      GPIO.add_event_detect(
22.04.23 04:34:13 (-0700)  gateway-config  RuntimeError: Failed to add edge detection
```

**Comment by shawaj** [comment]
@VeniceInventors please create a new issue if you still have something not working.

**Comment by shawaj** [comment]
Also @VeniceInventors FYI the issue you describe with the config container is usually because you haven't configured the config variables correctly.

Specifically the spi0-1cs one

**Comment by VeniceInventors** [comment]
> Also @VeniceInventors FYI the issue you describe with the config container is usually because you haven't configured the config variables correctly.
> 
> Specifically the spi0-1cs one

@shawaj  Thanks for the hint.

I didn't know about that variable, I'm using the nebra-helium-rak balena image 2.113.18, and the only variables I see that came pre-defined are:

FREQ -> N/A (Non Nebra Miner)
HELIUM_MINER_HEIGHT_URL -> https://api.helium.io/v1/blocks/height
VARIANT -> COMP-RAKHM

Is there a document somewhere which covers the spi0-1cs variable?
I don't want to take up your time if I can figure it out by reading some docs.

Should I create a new issue as well? Since it happens with the stock image, I'm guessing it happens to all the RAK users.

**Comment by shawaj** [comment]
@VeniceInventors where did you get that image from?

And how did you install on BalenaCloud?

Easiest way is using the deploy with Balena button here... https://github.com/NebraLtd/helium-rak#forking-fleet-installation-requires-updating-manually--setting-up-own-automations

**Comment by shawaj** [comment]
Also FYI you can see all the required variables here:
https://github.com/NebraLtd/helium-rak/blob/17ae057412f819180878a621d899bc0057df2416/balena.yml#L26-L43

**Comment by VeniceInventors** [comment]
> @VeniceInventors where did you get that image from?
> 
> And how did you install on BalenaCloud?
> 
> Easiest way is using the deploy with Balena button here... https://github.com/NebraLtd/helium-rak#forking-fleet-installation-requires-updating-manually--setting-up-own-automations
@shawaj 
That's how I deployed it, from the link above, downloaded the image from https://dashboard.balena-cloud.com/deploy?repoUrl=https://github.com/NebraLtd/helium-rak and flashed it, and added it to my balena hub fleet.

**Comment by VeniceInventors** [comment]
@shawaj 
I found the RESIN_HOST_CONFIG_dtoverlay: '"spi0-1cs","vc4-kms-v3d"' setting in the "Device Configuration" tab of balena.

It was set to "spi0-1cs","vc4-kms-v3d" for the fleet, but only to "vc4-kms-v3d" for the device itself. I don't know why as I didn't touch these variables.
I added the "spi0-1cs" and rebooted, and now the gateway-config is working as intended.

Thank you for your help!
